File: 20-valid-parentheses/valid-parentheses.java
// Timecomplexity : O[s] | Space complexity: O[s]  where s is length of stack
class Solution {
    public boolean isValid(String s) {
        if(s==null || s.length() == 0){
            return true;
        }

        // //declare and initialize the character array
        // char[] charArray = { '(',')','{','}','[',']' };

        Stack<Character> stck = new Stack<>();
        for(int i=0;i<s.length();i++){
            char c = s.charAt(i);
            if(c=='(')
                stck.push(')');

            else if(c=='{')
                stck.push('}');

            else if(c=='[')
                stck.push(']');

            else if (stck.isEmpty() || stck.pop() !=c)
                return false;
        }

        return stck.isEmpty();

        

    }


}

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        hashMap = {")" : "(", "]": "[", "}": "{"}
        stack = []

        for char in s:
            if char in hashMap:
                if stack and stack[-1] == hashMap[char]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(char)

        return not stack


















        # stack = []
        # hashMap = {")" : "(", "]":"[", "}":"{"}

        # for char in s:
        #     if char in hashMap:
        #         if stack and stack[-1] == hashMap[char]:
        #             stack.pop()
        #         else:
        #             return False
        #     else:
        #         stack.append(char)

        # return True if not stack else False


# Time Complexity: O(N)
# Space Complexity: O(N)
        
 
        